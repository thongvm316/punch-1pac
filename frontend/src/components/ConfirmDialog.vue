<template>
  <div
    class="modal"
    :class="{ active: open }"
  >
    <span
      class="modal-overlay"
      @click="toggle"
    />
    <div class="modal-container">
      <div class="modal-header">
        <span
          class="btn btn-clear float-right"
          @click="toggle"
        />
        <h3 class="modal-title">
          {{ title }}
        </h3>
      </div>
      <div class="modal-body">
        <div class="content">
          <slot />
        </div>
      </div>
      <div class="modal-footer">
        <slot name="confirmBtn">
          <button
            type="button"
            class="btn btn-error"
            @click="confirm"
          >
            {{ $t('confirmDialog.yes') }}
          </button>
        </slot>
        <button
          type="button"
          class="btn"
          @click="toggle"
        >
          {{ $t('confirmDialog.no') }}
        </button>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'ConfirmDialog',

  props: {
    title: {
      type: String,
      required: true
    },
    modalOpen: {
      type: Boolean,
      required: true
    },
    deleteObject: Function,
    objectId: Number
  },

  data() {
    return {
      open: this.modalOpen
    }
  },

  watch: {
    modalOpen() {
      this.open = this.modalOpen
    }
  },

  methods: {
    toggle() {
      this.open = !this.open
      this.$emit('update:modalOpen', this.open)
    },

    confirm() {
      this.deleteObject(this.objectId).then(() => {
        this.open = false
        this.$emit('update:modalOpen', this.open)
      })
    }
  }
}
</script>
