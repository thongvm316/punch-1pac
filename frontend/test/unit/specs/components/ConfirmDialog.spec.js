import { shallowMount } from '@vue/test-utils'

import localVue from '../../supports/local-vue'

import i18n from '@/locale'
import ConfirmDialog from '@/components/ConfirmDialog'

const propsData = {
  title: 'Confirm Dialog',
  modalOpen: false
}

const toggle = jest.spyOn(ConfirmDialog.methods, 'toggle')
const methods = {
  toggle,
  confirm: jest.fn()
}

const scopedSlots = {
  default: '<p class="default-scoped">Scope slot</p>',
  confirmBtn: '<p class="confirm-btn-scoped">Confirm Btn Slot</p>'
}

describe('ConfirmDialog.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(ConfirmDialog, {
      i18n,
      propsData,
      methods,
      scopedSlots,
      localVue
    })
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when ConfirmDialog was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('should render correct props data', () => {
      const { title, modalOpen } = wrapper.vm.$props

      expect(title).toEqual('Confirm Dialog')
      expect(wrapper.find('.modal-title').text()).toEqual('Confirm Dialog')
      expect(modalOpen).toEqual(false)
      expect(wrapper.find('.modal').classes()).not.toContain('active')
    })

    it('should render slot', () => {
      expect(wrapper.find('p.default-scoped').exists()).toBe(true)
      expect(wrapper.find('.modal-footer .confirm-btn-scoped').exists()).toBe(true)
    })
  })

  describe('when validation propsData', () => {
    it('should validate prop data', () => {
      const { title, modalOpen, deleteObject, objectId } = wrapper.vm.$options.props

      expect(title.required).toBeTruthy()
      expect(title.type).toBe(String)
      expect(modalOpen.required).toBeTruthy()
      expect(modalOpen.type).toBe(Boolean)
      expect(deleteObject.type).toBe(Function)
      expect(objectId.type).toBe(Number)
    })
  })

  describe('when methods confirm dialog', () => {
    beforeEach(async () => {
      wrapper.setProps({ modalOpen: true })
      await wrapper.vm.$nextTick()
    })

    it('should toggle method is called', () => {
      wrapper.find('.modal-overlay').trigger('click')

      expect(toggle).toHaveBeenCalled()
      expect(wrapper.vm.open).toBe(false)
      expect(wrapper.emitted('update:modalOpen')).toBeTruthy()
      expect(wrapper.emitted('update:modalOpen')).toHaveLength(1)
      expect(wrapper.emitted('update:modalOpen')[0]).toEqual([false])
    })
  })

  describe('when watcher modalOpen', () => {
    it('should open dialog when props change', async () => {
      expect(wrapper.vm.open).toBe(false)

      wrapper.setProps({ modalOpen: true })
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.open).toBe(true)
    })
  })
})
