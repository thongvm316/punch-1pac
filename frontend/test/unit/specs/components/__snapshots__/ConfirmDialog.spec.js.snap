// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog.vue when ConfirmDialog was mounted should render correctly 1`] = `
<div class="modal"><span class="modal-overlay"></span>
  <div class="modal-container">
    <div class="modal-header"><span class="btn btn-clear float-right"></span>
      <h3 class="modal-title">
        Confirm Dialog
      </h3>
    </div>
    <div class="modal-body">
      <div class="content">
        <p class="default-scoped">Scope slot</p>
      </div>
    </div>
    <div class="modal-footer"><button type="button" class="btn btn-error">
        Yes
      </button> <button type="button" class="btn">
        No
      </button></div>
  </div>
</div>
`;

exports[`ConfirmDialog.vue when watch modalOpen should active modal 1`] = `
<div class="modal active"><span class="modal-overlay"></span>
  <div class="modal-container">
    <div class="modal-header"><span class="btn btn-clear float-right"></span>
      <h3 class="modal-title">
        Confirm Dialog
      </h3>
    </div>
    <div class="modal-body">
      <div class="content">
        <p class="default-scoped">Scope slot</p>
      </div>
    </div>
    <div class="modal-footer"><button type="button" class="btn btn-error">
        Yes
      </button> <button type="button" class="btn">
        No
      </button></div>
  </div>
</div>
`;
